extensions/pasteboard/watcher.m

```diff
@@ -0,0 +1,316 @@
+    [skin checkArgs:LS_TFUNCTION, LS_TSTRING | LS_TOPTIONAL, LS_TBREAK];
+
+    NSString *pbName = [skin toNSObjectAtIndex:2];
+    
+    lua_pushvalue(L, 1);
+    int callbackRef = [skin luaRef:refTable];
+
+    // Create the timer object:
+    HSPasteboardTimer *timer = createHSPasteboardTimer(callbackRef, pbName);
+    
+    // Start the timer:
+    [timer start];
+
+    // Wire up the timer object to Lua:
+    void **userData = lua_newuserdata(L, sizeof(HSPasteboardTimer*));
+    *userData = (__bridge_retained void*)timer;
+    luaL_getmetatable(L, USERDATA_TAG);
+    lua_setmetatable(L, -2);
+
+    return 1;
+}
+
+/// hs.pasteboard.watcher:start() -> timer
+/// Method
+/// Starts an `hs.pasteboard.watcher` object
+///
+/// Parameters:
+///  * None
+///
+/// Returns:
+///  * The `hs.pasteboard.watcher` object
+static int pasteboardwatcher_start(lua_State* L) {
+    LuaSkin *skin = [LuaSkin sharedWithState:L];
+    [skin checkArgs:LS_TUSERDATA, USERDATA_TAG, LS_TBREAK];
+
+    HSPasteboardTimer* timer = get_objectFromUserdata(__bridge HSPasteboardTimer, L, 1, USERDATA_TAG);
+    lua_settop(L, 1);
+    
+    // Start the timer:
+    [timer start];
+
+    return 1;
+}
+
+/// hs.pasteboard.watcher:running() -> boolean
+/// Method
+/// Returns a boolean indicating whether or not the Pasteboard Watcher is currently running.
+///
+/// Parameters:
+///  * None
+///
+/// Returns:
+///  * A boolean value indicating whether or not the timer is currently running.
+static int pasteboardwatcher_running(lua_State* L) {
+    LuaSkin *skin = [LuaSkin sharedWithState:L];
+    [skin checkArgs:LS_TUSERDATA, USERDATA_TAG, LS_TBREAK];
+    HSPasteboardTimer *timer = get_objectFromUserdata(__bridge HSPasteboardTimer, L, 1, USERDATA_TAG);
+
+    lua_pushboolean(L, timer.isRunning);
+
+    return 1;
+}
+
+/// hs.pasteboard.watcher:stop() -> timer
+/// Method
+/// Stops an `hs.pasteboard.watcher` object
+///
+/// Parameters:
+///  * None
+///
+/// Returns:
+///  * The `hs.pasteboard.watcher` object
+static int pasteboardwatcher_stop(lua_State* L) {
+    LuaSkin *skin = [LuaSkin sharedWithState:L];
+    [skin checkArgs:LS_TUSERDATA, USERDATA_TAG, LS_TBREAK];
+    HSPasteboardTimer *timer = get_objectFromUserdata(__bridge HSPasteboardTimer, L, 1, USERDATA_TAG);
+    lua_settop(L, 1);
+
+    // Stop the timer:
+    [timer stop];
+
+    return 1;
+}
+
+static int pasteboardwatcher_gc(lua_State* L) {
+    LuaSkin *skin = [LuaSkin sharedWithState:L];
+    HSPasteboardTimer *timer = get_objectFromUserdata(__bridge_transfer HSPasteboardTimer, L, 1, USERDATA_TAG);
+
+    if (timer) {
+        [timer stop];
+        timer.fnRef = [skin luaUnref:refTable ref:timer.fnRef];
+        timer.t = nil;
+        timer.pbName = nil;
+        timer = nil;
+    }
+    
+    // Remove the Metatable so future use of the variable in Lua won't think its valid
+    lua_pushnil(L);
+    lua_setmetatable(L, 1);
+
+    return 0;
+}
+
+static int meta_gc(lua_State* __unused L) {
```

Comment:
By not invalidating `sharedPasteboardTimer` and setting it to nil here in the `meta_gc` function, you do realize that your very first watcher is never released by ARC (its notification watcher and callback function ref *are* cleared by it's lua __gc function, though, so it isn't dangling within `lua`) even after a reload of Hammerspoon/CommandPost and will still be running every half second, even if no watchers are created after the reload.

Minor impact, but I'd be more comfortable with something like this here (untested, but I think this should be sufficient):

~~~objc
    if (sharedPasteboardTimer) {
        [sharedPasteboardTimer invalidate] ;
        sharedPasteboardTimer = nil ;
    }
~~~

Reply:
Good point, thanks!